### DISCUSSION TITLE
Configuration and Offline Usage of MCP Server

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-theia/theia#14685] [Q&A]
Hi,
I’m currently setting up an MCP server. for example @modelcontextprotocol/server-filesystem, After configuring and starting it, I need to build my app to run in an offline environment. I noticed in the [MCP introduction](https://eclipsesource.com/blogs/2024/12/19/theia-ide-and-theia-ai-support-mcp/) that the required Node packages are downloaded during the first startup. I’d like to know where these MCP Node packages are stored so I can manually download them and set up additional MCP servers for offline use.
![image](https://github.com/user-attachments/assets/3c8f6b58-7dba-4cde-9ec9-fd3eabcb0627)

**Comment by JonasHelming** [comment]
We simply call npx -y to run the server. 

To ensure that the npx -y @modelcontextprotocol/server-filesystem command works without an internet connection, you need to pre-download the required package and cache it locally. 

Run the following command to globally install the package on your system:

`npm install -g @modelcontextprotocol/server-filesystem`

This will make the package available globally on your machine, and npx will find it without needing to fetch it from the registry.

**Reply by wss29** [reply]
hi @JonasHelming 
  A more effective approach I've found is to manually download the @modelcontextprotocol/server-filesystem package and place it directly into the node_modules directory of Theia. This ensures that every time I start the server, it launches quickly without issues.

While using npm install -g @modelcontextprotocol/server-filesystem is an option, I've noticed that when starting the MCP server, it sometimes fails to connect or is slow even when the connection is successful.

Additionally, attempting to install the package with yarn add @modelcontextprotocol/server-filesystem consistently results in errors. As a workaround, I resort to manually downloading and copying the package to the node_modules folder.

**Comment by sdirix** [comment]
The `npx` command is just a convenient way of downloading and starting the respective MCP server, it's not the only way. In the end our MCP integration does not care about what command is executed, just that the started program acts like a MCP server.

For example to run the reference servers offline, you can build them from scratch and then point Node to their script, [just like they declare in their reference server package.jsons](https://github.com/modelcontextprotocol/servers/blob/69bfe74bebf9101f233343fcafe543f94b94d2d1/src/memory/package.json#L11), e.g. the command `node <absolute-path-on-your-fileystem-to-git-repo>/src/memory/dist/index.js` should work the same without requiring internet access (besides the first clone / build of the repo) nor installing anything global on your machine.
